### ISSUE TITLE
Inference: ability to parallelize (multi-gpu) inference on a single image

### ISSUE THREAD

**Issue by remtav** [NRCan/geo-deep-learning#182]
Currently GDL doesn't have the capacity to infer a single image on multiple GPUs (i.e. parallel forward pass). 

The issue becomes urgent the moment we infer on very large images, which has been the case lately at NRCan with images over 48k x 48k pixels (BC_16).

For AI as a service, this feature will be either helpful or necessery to offer a better "user experience".

Some suggestions for implementation:
1. Read image --> split --> save into temporary thdf5s --> use dataloaders the same way we use them for training, allowing inference on 2+ gpus.
2. Read image and keep in memory until inference is finished --> feed bits of it (not sure how exactly) to our model. Is there a dataloader that can deal with this more "dynamic" approach? This second approach needs some further thinking but could help infer more quickly than approach #1.

Any thoughts?

**Comment by ymoisan** [comment]
I would implement DataParallel for inference and no more for now.  I think the problem is more one of maximizing inference speed rather than being able to handle huge images.  Maybe we should have a service to extract "sub-images" from huge COGs and supply those sub-images for inference on (potentially) more than one GPU.  That is, until we can actually stream the COGs in the inference engine dynamically.

**Comment by remtav** [comment]
@ymoisan naturally, we will need DataParallel. The problem is Dataparallel intimately works with a Pytorch Dataloader. We don't use a Dataloader currently at inference. Therefore, I send you back to my two ideas. We'll need to elaborate on the simplest/best solution for fast/efficient inference.

**Comment by mpelchat04** [comment]
Consideration:  
* How does this conflicts/interact with test-time augmentations (which takes up quite a bit of GPU memory, same as batching the process)?

**Comment by ymoisan** [comment]
Another consideration:
- The more GPUs are required on HPC, the more the chances of delaying job execution

**Comment by remtav** [comment]
> Consideration:
> 
> * How does this conflicts/interact with test-time augmentations (which takes up quite a bit of GPU memory, same as batching the process)?

I'm pretty sure TTA could be used easily in a multi-GPU execution (managed by pytorch/pytorch-lightning) with DataParallel or DistributedDataParallel). Using more than 1 gpu could potentially speedup the inference process when TTA is used. The simplest implementation would be for GPUs to process different tiles from the source imagery and be able to write to a common numpy array. Tests would be needed as this is not Pytorch's default use case.

**Comment by remtav** [comment]
> Another consideration:
> 
> * The more GPUs are required on HPC, the more the chances of delaying job execution

Indeed. In the context or deploying on HPC, we would greatly benefit from having GPUs with 32gb ram rather than 16gb for inference. The pros of bigger GPU would definitely outweight the pros of multi-gpu inference in our current setup, since a bigger gpu needs no further development effort...

**Comment by remtav** [comment]
Just tested gpu inference using torchgeo templates and pytorch lightning (PL). I had in mind to make it work with a single gpu in a PL context. Then, porting to multi-gpu would have been just a question of settings gpus to more than 1 (PL default to ddp-spawn). 

It seems multi-gpu feature in PL is mostly aimed at training/validation. See:
https://stackoverflow.com/questions/65807601/output-prediction-of-pytorch-lightning-model

For example, PL-Trainer's auto_scale_batch_size is really only aimed at training. Here's the error I got when trying for inference:
`pytorch_lightning.utilities.exceptions.MisconfigurationException: No `training_step()` method defined. Lightning `Trainer` expects as minimum a `training_step()`, `train_dataloader()` and `configure_optimizers()` to be defined.
`

Just for the record here's essentially the code version for my best trial (see inference_segmentation.py in my 222-stac-item-input branch for the rest):

```
    checkpoint, hparams = gdl2pl_checkpoint(checkpoint)
    [...]
    task = InferenceTask(checkpoint, **hparams)
    trainer_args = {}
    trainer_args["default_root_dir"] = dm.root_dir

    dm.batch_size = task.batch_size = 4
    trainer = Trainer(auto_scale_batch_size=True, gpus=0, precision=16, **trainer_args)
    trainer.tune(task)
    predictions = trainer.predict(model=task, datamodule=dm, return_predictions=True, ckpt_path=checkpoint)
```

In any case, I'm closing this issue as multi-gpu inference is not really a need we have. We'd rather run multiple inferences at once on different images and use one gpu per image, then using multigpu approach on a single image.
